server/s3_motr_writer.cc

```diff
@@ -722,11 +722,149 @@ int S3MotrWiter::get_op_ret_code_for_delete_op(int index) {
+    size_t &starting_checksum_buf_idx, bool &calculated_chksum_at_unit_boundary,
+    bool reset_initial_buffers, bool is_called_for_unaligned_buffers,
+    bool is_finalize_call) {
+  int rc = -1;
+  assert(rw_ctx != NULL);
+  s3_log(S3_LOG_DEBUG, "", "%s Entry", __func__);
+
+  // Calling init once
+  if ((saved_last_index == 0) || (initial_buffers_part_write)) {
+    // If the write is with offset as 0 or multipart part
+    // upload (offset wont be 0 for R1) for initial write
+    // we need to call init
+    s3_checksum_flag |= S3_FIRST_UNIT;
+
+    if (reset_initial_buffers) {
+      initial_buffers_part_write = false;
+    }
+  }
+
+  s3_log(S3_LOG_DEBUG, request_id,
+         "Calculating checksum (%zu), "
+         "seed_offset(%zu) chksum_buf_idx(%zu)\n",
+         size_in_current_write, saved_last_index, chksum_buf_idx);
+
+  if (is_finalize_call) {
+    // Update (with actual data length) + Finalize without seed
+    rc = s3_md5crypt->s3_calculate_unaligned_buffs_pi(rw_ctx, s3_checksum_flag);
+    if (rc != 0) {
+      s3_log(S3_LOG_ERROR, request_id,
+             "motr api motr_client_calculate_pi failed with return code %d\n",
+             rc);
+      // Should we stop here? what should we do?
+    }
+    assert(rc == 0);
+  } else {
+    if (!(s3_checksum_flag & S3_SEED_UNIT) && is_s3_write_di_check_enabled) {
+      s3_checksum_flag |= S3_SEED_UNIT;
+    }
+
+    rc = s3_md5crypt->s3_calculate_unit_pi(rw_ctx, chksum_buf_idx++,
+                                           saved_last_index, oid_list[0],
+                                           s3_checksum_flag);
+    if (rc != 0) {
+      s3_log(S3_LOG_ERROR, request_id,
+             "motr api motr_client_calculate_pi failed with return code %d\n",
+             rc);
+      // Should we stop here? what should we do?
+    }
+    assert(rc == 0);
+  }
+
+  if (!is_called_for_unaligned_buffers) {
+    saved_last_index = last_index;
+    starting_checksum_buf_idx = 0;
+    unaligned_buf_idx_offset = buf_idx;
+
+    if (s3_checksum_flag & S3_FIRST_UNIT) {
+      s3_checksum_flag &= ~S3_FIRST_UNIT;
+    }
+
+    calculated_chksum_at_unit_boundary = true;
+  }
+
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+void S3MotrWiter::find_and_allocate_placeholder_for_data_alignment(
+    size_t &buf_idx, size_t &motr_buf_count) {
+  s3_log(S3_LOG_DEBUG, "", "%s Entry", __func__);
+
+  if (buf_idx < motr_buf_count) {
+    // Allocate place_holder_for_last_unit only if its required.
+    // Its required when writing last block of object.
+    unit_size_for_place_holder =
+        g_option_instance->get_libevent_pool_buffer_size();
+    reset_buffers_if_any(unit_size_for_place_holder);
+    place_holder_for_last_unit =
+        (void *)S3MempoolManager::get_instance()->get_buffer_for_unit_size(
+            unit_size_for_place_holder);
+  }
+
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+void S3MotrWiter::align_data_to_motr_unit_size(
```

Comment:
Please write function headers / Comments about the functionalities about these functions